File: game-engine/src/test/java/org/luxons/sevenwonders/game/cards/RequirementsTest.java
package org.luxons.sevenwonders.game.cards;

import java.util.Arrays;
import java.util.Collections;

import org.junit.Test;
import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.runner.RunWith;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.resources.Provider;
import org.luxons.sevenwonders.game.resources.ResourceTransactions;
import org.luxons.sevenwonders.game.resources.ResourceType;
import org.luxons.sevenwonders.game.resources.Resources;
import org.luxons.sevenwonders.game.test.TestUtilsKt;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.junit.Assume.assumeTrue;

@RunWith(Theories.class)
public class RequirementsTest {

    @DataPoints
    public static int[] goldAmounts() {
        return new int[] {0, 1, 2, 5};
    }

    @DataPoints
    public static ResourceType[] resourceTypes() {
        return ResourceType.values();
    }

    @Test
    public void getResources_emptyAfterInit() {
        Requirements requirements = new Requirements();
        assertTrue(requirements.getResources().isEmpty());
    }

    @Test
    public void setResources_success() {
        Requirements requirements = new Requirements();
        Resources resources = new Resources();
        requirements.setResources(resources);
        assertSame(resources, requirements.getResources());
    }

    @Theory
    public void goldRequirement(int boardGold, int requiredGold) {
        Requirements requirements = new Requirements();
        requirements.setGold(requiredGold);

        Board board = TestUtilsKt.testBoard(ResourceType.CLAY, boardGold);
        Table table = new Table(Collections.singletonList(board));

        assertEquals(boardGold >= requiredGold, requirements.areMetWithoutNeighboursBy(board));
        assertEquals(boardGold >= requiredGold, requirements.areMetWithHelpBy(board, new ResourceTransactions()));
        assertEquals(boardGold >= requiredGold, requirements.areMetBy(table, 0));
    }

    @Theory
    public void resourceRequirement_initialResource(ResourceType initialResource, ResourceType requiredResource) {
        Requirements requirements = TestUtilsKt.createRequirements(requiredResource);

        Board board = TestUtilsKt.testBoard(initialResource, 0);
        Table table = new Table(Collections.singletonList(board));

        assertEquals(initialResource == requiredResource, requirements.areMetWithoutNeighboursBy(board));
        assertEquals(initialResource == requiredResource,
                requirements.areMetWithHelpBy(board, new ResourceTransactions()));

        if (initialResource == requiredResource) {
            assertTrue(requirements.areMetBy(table, 0));
        }
    }

    @Theory
    public void resourceRequirement_ownProduction(ResourceType initialResource, ResourceType producedResource,
            ResourceType requiredResource) {
        assumeTrue(initialResource != requiredResource);

        Requirements requirements = TestUtilsKt.createRequirements(requiredResource);

        Board board = TestUtilsKt.testBoard(initialResource, 0);
        board.getProduction().addFixedResource(producedResource, 1);
        Table table = new Table(Collections.singletonList(board));

        assertEquals(producedResource == requiredResource, requirements.areMetWithoutNeighboursBy(board));
        assertEquals(producedResource == requiredResource,
                requirements.areMetWithHelpBy(board, new ResourceTransactions()));

        if (producedResource == requiredResource) {
            assertTrue(requirements.areMetBy(table, 0));
        }
    }

    @Theory
    public void resourceRequirement_boughtResource(ResourceType initialResource, ResourceType boughtResource,
            ResourceType requiredResource) {
        assumeTrue(initialResource != requiredResource);

        Requirements requirements = TestUtilsKt.createRequirements(requiredResource);

        Board board = TestUtilsKt.testBoard(initialResource, 2);
        Board neighbourBoard = TestUtilsKt.testBoard(initialResource, 0);
        neighbourBoard.getPublicProduction().addFixedResource(boughtResource, 1);
        Table table = new Table(Arrays.asList(board, neighbourBoard));

        ResourceTransactions resources = TestUtilsKt.createTransactions(Provider.RIGHT_PLAYER, boughtResource);

        assertFalse(requirements.areMetWithoutNeighboursBy(board));
        assertEquals(boughtResource == requiredResource, requirements.areMetWithHelpBy(board, resources));

        if (boughtResource == requiredResource) {
            assertTrue(requirements.areMetBy(table, 0));
        }
    }

    @Theory
    public void pay_boughtResource(ResourceType initialResource, ResourceType requiredResource) {
        assumeTrue(initialResource != requiredResource);

        Requirements requirements = TestUtilsKt.createRequirements(requiredResource);

        Board board = TestUtilsKt.testBoard(initialResource, 2);
        Board neighbourBoard = TestUtilsKt.testBoard(requiredResource, 0);
        Table table = new Table(Arrays.asList(board, neighbourBoard));

        ResourceTransactions transactions = TestUtilsKt.createTransactions(Provider.RIGHT_PLAYER,
                requiredResource);

        assertFalse(requirements.areMetWithoutNeighboursBy(board));
        assertTrue(requirements.areMetWithHelpBy(board, transactions));
        assertTrue(requirements.areMetBy(table, 0));

        requirements.pay(table, 0, transactions);

        assertEquals(0, board.getGold());
        assertEquals(2, neighbourBoard.getGold());
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/cards/RequirementsTest.kt
package org.luxons.sevenwonders.game.cards

import org.junit.Assume.assumeTrue
import org.junit.Test
import org.junit.experimental.theories.DataPoints
import org.junit.experimental.theories.Theories
import org.junit.experimental.theories.Theory
import org.junit.runner.RunWith
import org.luxons.sevenwonders.game.SimplePlayer
import org.luxons.sevenwonders.game.boards.Table
import org.luxons.sevenwonders.game.resources.Provider
import org.luxons.sevenwonders.game.resources.ResourceType
import org.luxons.sevenwonders.game.resources.emptyResources
import org.luxons.sevenwonders.game.resources.noTransactions
import org.luxons.sevenwonders.game.test.createRequirements
import org.luxons.sevenwonders.game.test.createTransactions
import org.luxons.sevenwonders.game.test.singleBoardPlayer
import org.luxons.sevenwonders.game.test.testBoard
import kotlin.test.assertEquals
import kotlin.test.assertSame
import kotlin.test.assertTrue

@RunWith(Theories::class)
class RequirementsTest {

    @Test
    fun getResources_emptyAfterInit() {
        val (_, resources) = Requirements()
        assertTrue(resources.isEmpty())
    }

    @Test
    fun setResources_success() {
        val resources = emptyResources()
        val requirements = Requirements(0, resources)
        assertSame(resources, requirements.resources)
    }

    @Theory
    fun goldRequirement(boardGold: Int, requiredGold: Int) {
        val requirements = Requirements(requiredGold)

        val board = testBoard(ResourceType.CLAY, boardGold)
        val player = singleBoardPlayer(board)

        assertEquals(boardGold >= requiredGold, requirements.areMetWithHelpBy(board, noTransactions()))

        val satisfaction = requirements.computeSatisfaction(player)
        if (boardGold >= requiredGold) {
            if (requiredGold == 0) {
                assertEquals(RequirementsSatisfaction.noRequirements(), satisfaction)
            } else {
                assertEquals(RequirementsSatisfaction.enoughGold(requiredGold), satisfaction)
            }
        } else {
            assertEquals(RequirementsSatisfaction.missingRequiredGold(requiredGold), satisfaction)
        }
    }

    @Theory
    fun resourceRequirement_initialResource(initialResource: ResourceType, requiredResource: ResourceType) {
        val requirements = createRequirements(requiredResource)

        val board = testBoard(initialResource, 0)
        val player = singleBoardPlayer(board)

        assertEquals(initialResource == requiredResource, requirements.areMetWithHelpBy(board, noTransactions()))

        if (initialResource == requiredResource) {
            val satisfaction = requirements.computeSatisfaction(player)
            assertEquals(RequirementsSatisfaction.enoughResources(), satisfaction)
        }
    }

    @Theory
    fun resourceRequirement_ownProduction(
        initialResource: ResourceType,
        producedResource: ResourceType,
        requiredResource: ResourceType
    ) {
        assumeTrue(initialResource != requiredResource)

        val requirements = createRequirements(requiredResource)

        val board = testBoard(initialResource, 0)
        board.production.addFixedResource(producedResource, 1)
        val player = singleBoardPlayer(board)

        assertEquals(producedResource == requiredResource, requirements.areMetWithHelpBy(board, noTransactions()))

        if (producedResource == requiredResource) {
            val satisfaction = requirements.computeSatisfaction(player)
            assertEquals(RequirementsSatisfaction.enoughResources(), satisfaction)
        }
    }

    @Theory
    fun resourceRequirement_boughtResource(
        initialResource: ResourceType,
        boughtResource: ResourceType,
        requiredResource: ResourceType
    ) {
        assumeTrue(initialResource != requiredResource)

        val requirements = createRequirements(requiredResource)

        val board = testBoard(initialResource, 2)
        val neighbourBoard = testBoard(initialResource, 0)
        neighbourBoard.publicProduction.addFixedResource(boughtResource, 1)
        val table = Table(listOf(board, neighbourBoard))
        val player = SimplePlayer(0, table)

        val resources = createTransactions(Provider.RIGHT_PLAYER, boughtResource)

        val neighbourHasResource = boughtResource == requiredResource
        assertEquals(neighbourHasResource, requirements.areMetWithHelpBy(board, resources))

        val satisfaction = requirements.computeSatisfaction(player)
        if (neighbourHasResource) {
            val transactions = setOf(
                createTransactions(Provider.LEFT_PLAYER, requiredResource),
                createTransactions(Provider.RIGHT_PLAYER, requiredResource)
            )
            assertEquals(RequirementsSatisfaction.metWithHelp(2, transactions), satisfaction)
        } else {
            assertEquals(RequirementsSatisfaction.resourcesUnavailable(), satisfaction)
        }
    }

    @Theory
    fun pay_boughtResource(initialResource: ResourceType, requiredResource: ResourceType) {
        assumeTrue(initialResource != requiredResource)

        val requirements = createRequirements(requiredResource)

        val board = testBoard(initialResource, 2)
        val neighbourBoard = testBoard(requiredResource, 0)
        val table = Table(listOf(board, neighbourBoard))
        val player = SimplePlayer(0, table)

        val transactions = createTransactions(Provider.RIGHT_PLAYER, requiredResource)

        assertTrue(requirements.areMetWithHelpBy(board, transactions))
        assertTrue(requirements.computeSatisfaction(player).satisfied)

        requirements.pay(player, transactions)

        assertEquals(0, board.gold)
        assertEquals(2, neighbourBoard.gold)
    }

    companion object {

        @JvmStatic
        @DataPoints
        fun goldAmounts(): IntArray = intArrayOf(0, 1, 2, 5)

        @JvmStatic
        @DataPoints
        fun resourceTypes(): Array<ResourceType> = ResourceType.values()
    }
}
